### ISSUE TITLE
Tooltip is placed off-screen

### ISSUE THREAD

**Issue by bobrik** [grafana/grafana#14134]
### What Grafana version are you using?

5.3.0

### What OS are you running grafana on?

Chrome on macOS Mojave

### What did you do?

Hovered a graph to see a tooltip.

### What was the expected result?

Tooltip is placed where it can be seen in its entirety.

### What happened instead?

This:

![image](https://user-images.githubusercontent.com/89186/48741499-37869980-ec10-11e8-9218-31794524e0ad.png)

When I hover upper graph below 40%, this happens, otherwise it's good:

![image](https://user-images.githubusercontent.com/89186/48741542-5be27600-ec10-11e8-8ca6-eb4a3d73e04d.png)

**Comment by Ijin08** [comment]
Confirmed bug. Able to reproduce it on MacBook Pro screen. Works fine on big screen.

**Comment by roidelapluie** [comment]
A problem especially on mobile :)

**Comment by ushuz** [comment]
I think this is the related code:

https://github.com/grafana/grafana/blob/2325c5dbedd231507179a9db5af3ff8518354e09/public/app/core/jquery_extended.ts#L45-L46

I propose the following fix, makes sure `$tooltip`'s `top` never goes below zero, so it always remains in viewport

```diff
// during my experiment, I noticed that the `x` received by the function 
// doesn't reflect actual pointer position, it has a slight offset around 
// 20px, so we have to mitigate this offset, otherwise the tooltip will 
// overlap with pointer and keep blinking under some circumstances after 
// we cap the css `top` value >= 0
//
// I'm not sure this is a widespread problem or not, but this line is 
// needed at least for me
+x -= 20; 
$tooltip.css('left', x + opts.offset + width > $win.width() ? x - opts.offset - width : x + opts.offset); 
-$tooltip.css('top', y + opts.offset + height > $win.height() ? y - opts.offset - height : y + opts.offset); 
+$tooltip.css('top', Math.max(0, y + opts.offset + height > $win.height() ? y - opts.offset - height : y + opts.offset));
```

For anyone who wants a immediate fix and uses docker to run grafana, you can try patching `app.js` inside container like this, **you may need to tweak the command, use at your own risk**

```bash
sed -i 's/o.css("top",n+a.offset+s>D.height()?n-a.offset-s:n+a.offset)/o.css("top",Math.max(0,n+a.offset+s>D.height()?n-a.offset-s:n+a.offset))/g' /usr/share/grafana/public/build/app.dfabdd44b3be44288eac.js
sed -i 's/),o.css("left",e+/),e-=20,o.css("left",e+/g' /usr/share/grafana/public/build/app.dfabdd44b3be44288eac.js
```

**Comment by ushuz** [comment]
It seems that #20874 fixed this and released in 6.5.2.
https://github.com/grafana/grafana/blob/master/CHANGELOG.md#652-2019-12-11

**Comment by msober** [comment]
Hi, We have figured out the problem. when solve the problem of preventing tooltip displayed outside of window in issure #20874, add releated code in [jquery_extended.ts](https://github.com/grafana/grafana/blob/master/public/app/core/jquery_extended.ts)

```javascript
$tooltip.css('left',left>0?left:0);
$tooltip.css('top',top>0?top:0);
```

however, in [graph_tooltip.ts](https://github.com/grafana/grafana/blob/master/public/app/plugins/panel/graph/graph_tooltip.ts) , the renderAndShow function of tooltip div add 20px to the position of x axis

```javascript
$tooltip.html(innerHtml).place_tt(pos.pageX+20,pos.pageY);
```

so the width param  include 20px, but the offset param is 0, the result  of left calculation is inaccurate.

```javascript
const left=x+opts.offset+width>$win.width()!?x-opts.offset-width:x+opts.offset;
```

In some condition, **this can cause tooltip div display under the mouse arrow, then tooltip div trigger onMouseLeave event that has bind to it, after onMouseLeave the tooltip div removed.**

```javascript
onMouseLeave={() => {
  this.setState({ isTooltipVisible: false });
}}
```

**now we fix problem in this way:** 

In renderAndShow function we use { offset: 20} as the third param, and remove pos.pageX+20. like:

```javascript
$tooltip.html(innerHtml).place_tt(pos.pageX, pos.pageY, { offset: 20 });
```

Howerver, that is not a perfet way, both x and y  add 20, so we suggest place_tt to support offsetX, offsetY param like:
```javascript
$tooltip.html(innerHtml).place_tt(pos.pageX, pos.pageY, { offsetX: 20, offsetY:0 });
```

**If necessary, we can sent a pull request to fix this bug.**
